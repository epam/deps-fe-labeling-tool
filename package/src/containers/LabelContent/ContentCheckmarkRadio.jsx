import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateLabels } from '@/actions/markup'
import { RadioGroup } from '@/components/Radio'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { labelShape } from '@/models/Label'
import { Option } from '@/models/Option'
import { currentPageSelector } from '@/selectors/pagination'
import { RadioGroupWrapper } from './ContentCheckmarkRadio.styles'

const RESOURCE_TRUE_OPTION = 'True'
const RESOURCE_FALSE_OPTION = 'False'

const CheckmarkOption = {
  [RESOURCE_FALSE_OPTION]: false,
  [RESOURCE_TRUE_OPTION]: true
}

const options = Object.entries(CheckmarkOption).map(([text, value]) => new Option(value, text))

const ContentCheckmarkRadio = ({
  currentPage,
  label,
  updateLabels
}) => {
  const onContentChange = useCallback(
    (value) => {
      updateLabels(currentPage, [{
        ...label,
        content: value,
        confidence: CONFIDENCE_ON_MANUAL_CHANGE
      }])
    },
    [
      currentPage,
      label,
      updateLabels
    ])

  return (
    <RadioGroupWrapper>
      <RadioGroup
        options={options}
        onChange={onContentChange}
        value={label.content}
      />
    </RadioGroupWrapper>
  )
}

ContentCheckmarkRadio.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  updateLabels: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state)
})

const mapDispatchToProps = {
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ContentCheckmarkRadio)

export {
  ConnectedComponent as ContentCheckmarkRadio
}
