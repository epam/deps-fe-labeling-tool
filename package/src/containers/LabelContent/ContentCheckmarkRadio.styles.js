
import styled from 'styled-components'
import { COLORS } from '@/theme/theme.default'

const RadioGroupWrapper = styled.div`
  display: flex;
  align-items: center;
  padding: 0 0.8rem 0 1.2rem;
  border: 1px solid ${COLORS.GRAYSCALE_1};
  border-radius: 4px;
  height: 3.2rem;
  width: 100%;
  max-width: 100%;
  transition: all 0.3s;

  &:focus-within {
    border-color: ${COLORS.PRIMARY_2};
  }

`

export {
  RadioGroupWrapper
}
