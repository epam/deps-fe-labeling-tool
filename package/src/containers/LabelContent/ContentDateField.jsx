import React, { useCallback } from 'react'
import dayjs from 'dayjs'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateLabels } from '@/actions/markup'
import { CalendarIcon } from '@/components/Icons/CalendarIcon'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { debounce } from '@/hocs/debounce'
import { labelShape } from '@/models/Label'
import { currentPageSelector } from '@/selectors/pagination'
import {
  LOCALE_DATE_FORMAT,
  dayjsToString,
  stringToDayjs
} from '@/utils/dayjs'
import {
  InputWrapper,
  StyledDatePicker,
  StyledInput
} from './ContentDateField.styles'

const DebouncedInput = debounce({
  getValue: (e) => e.target.value
})(StyledInput)

const dateNow = dayjs()

const ContentDateField = ({
  currentPage,
  label,
  updateLabels
}) => {
  const handleLabelsUpdate = useCallback(
    (content) => {
      updateLabels(currentPage, [{
        ...label,
        content,
        confidence: CONFIDENCE_ON_MANUAL_CHANGE
      }])
    },
    [
      currentPage,
      label,
      updateLabels
    ])

  const handleDatePick = useCallback(
    (dayjsInstance) =>
      handleLabelsUpdate(dayjsToString(dayjsInstance, LOCALE_DATE_FORMAT)),
    [handleLabelsUpdate]
  )

  return (
    <>
      <InputWrapper>
        <DebouncedInput
          value={label.content}
          onChange={handleLabelsUpdate}
        />
        <StyledDatePicker
          allowClear={false}
          value={
            stringToDayjs(label.content) ||
            dateNow
          }
          onChange={handleDatePick}
          suffixIcon={<CalendarIcon />}
        />
      </InputWrapper>
    </>
  )
}

ContentDateField.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  updateLabels: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state)
})

const mapDispatchToProps = {
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ContentDateField)

export {
  ConnectedComponent as ContentDateField
}
