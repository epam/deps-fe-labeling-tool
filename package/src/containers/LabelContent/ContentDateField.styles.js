import styled from 'styled-components'
import { DatePicker } from '@/components/DatePicker'
import { Input } from '@/components/Input'
import { COLORS } from '@/theme/theme.default'

export const InputWrapper = styled.div`
  position: relative;
`

export const StyledInput = styled(Input)`
  padding: 0 3.2rem 0 1.2rem;
  height: 3.2rem;
  color: ${COLORS.GRAYSCALE_8};
  font-size: 1.4rem;
  font-style: normal;
  font-weight: 400;
  line-height: 2.2rem;
`

export const StyledDatePicker = styled(DatePicker)`
  padding: 0;
  position: absolute;
  top: 50%;
  right: 0.8rem;
  border: none;
  box-shadow: none;
  translate: 0 -50%;
  line-height: 0;

  & .ant-picker-suffix {
    margin: 0;
  }

  & .ant-picker-clear {
    right: 50%;
    transform: translate(50%, -50%);
  }

  & .ant-picker-input {
    cursor: pointer;

    & > input {
      display: none;
    }
  }
`
