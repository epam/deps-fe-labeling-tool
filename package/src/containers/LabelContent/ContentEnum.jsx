import React, { useCallback, useMemo } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateLabels } from '@/actions/markup'
import { Autocomplete } from '@/components/Autocomplete'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { labelShape } from '@/models/Label'
import { Option } from '@/models/Option'
import { fieldsSelector } from '@/selectors/model'
import { currentPageSelector } from '@/selectors/pagination'

const ContentEnum = ({
  currentPage,
  label,
  updateLabels,
  fields
}) => {
  const onContentChange = useCallback(
    (value) => {
      updateLabels(currentPage, [{
        ...label,
        content: value,
        confidence: CONFIDENCE_ON_MANUAL_CHANGE
      }])
    },
    [
      currentPage,
      label,
      updateLabels
    ])

  const enumOptions = useMemo(() => {
    const { fieldType, fieldMeta } = fields.find((field) => field.code === label.fieldCode)
    const { options } = fieldType === FieldType.LIST ? fieldMeta.baseTypeMeta : fieldMeta

    return options.map((value) => new Option(value))
  }, [fields, label.fieldCode])

  return (
    <Autocomplete
      onChange={onContentChange}
      value={label.content}
      options={enumOptions}
    />
  )
}

ContentEnum.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  updateLabels: PropTypes.func.isRequired,
  fields: PropTypes.arrayOf(fieldShape.isRequired)
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  fields: fieldsSelector(state)
})

const mapDispatchToProps = {
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ContentEnum)

export {
  ConnectedComponent as ContentEnum
}
