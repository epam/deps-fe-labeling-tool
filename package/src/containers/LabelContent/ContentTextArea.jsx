import React, { useMemo, useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateLabels } from '@/actions/markup'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { SpecialSymbols } from '@/containers/SpecialSymbols'
import { Feature } from '@/enums/Feature'
import { debounce } from '@/hocs/debounce'
import { labelShape } from '@/models/Label'
import { Settings, settingsShape } from '@/models/Settings'
import { currentPageSelector } from '@/selectors/pagination'
import { settingsSelector } from '@/selectors/settings'
import { TextAreaStyled } from './ContentTextArea.styles'

const DebouncedTextArea = debounce({
  getValue: (e) => e.target.value
})(TextAreaStyled)

const ContentTextArea = ({
  currentPage,
  label,
  settings,
  updateLabels
}) => {
  const areSymbolsAvailable = useMemo(
    () => (
      Settings.has(settings, Feature.SPECIAL_SYMBOLS)
    ),
    [settings]
  )

  const onContentChange = useCallback(
    (value) => {
      updateLabels(currentPage, [{
        ...label,
        content: value,
        confidence: CONFIDENCE_ON_MANUAL_CHANGE
      }])
    },
    [
      currentPage,
      label,
      updateLabels
    ])

  return (
    <>
      <DebouncedTextArea
        rows={2}
        value={label.content}
        onChange={onContentChange}
      />
      {
        areSymbolsAvailable && (
          <SpecialSymbols
            label={label}
          />
        )
      }
    </>
  )
}

ContentTextArea.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  settings: settingsShape.isRequired,
  updateLabels: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  settings: settingsSelector(state)
})

const mapDispatchToProps = {
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ContentTextArea)

export {
  ConnectedComponent as ContentTextArea
}
