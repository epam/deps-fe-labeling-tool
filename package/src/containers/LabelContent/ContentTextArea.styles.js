import styled from 'styled-components'
import { TextArea } from '@/components/TextArea'

const TextAreaStyled = styled(TextArea)`
  position: initial;
`

export {
  TextAreaStyled
}
