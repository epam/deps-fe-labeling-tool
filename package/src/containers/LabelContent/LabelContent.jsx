import React, { useMemo } from 'react'
import {
  LabelType,
  labelShape
} from '@/models/Label'
import { ContentCheckmarkRadio } from './ContentCheckmarkRadio'
import { ContentDateField } from './ContentDateField'
import { ContentEnum } from './ContentEnum'
import { ContentTextArea } from './ContentTextArea'

const MAP_LABEL_TYPE_TO_CONTENT_COMPONENT = {
  [LabelType.STRING]: ContentTextArea,
  [LabelType.CHECKMARK]: ContentCheckmarkRadio,
  [LabelType.ENUM]: ContentEnum,
  [LabelType.DATE]: ContentDateField
}

const LabelContent = ({
  label
}) => {
  const Content = useMemo(
    () => MAP_LABEL_TYPE_TO_CONTENT_COMPONENT[label.type] ?? ContentTextArea,
    [label.type]
  )

  return (
    <Content label={label} />
  )
}

LabelContent.propTypes = {
  label: labelShape.isRequired
}

export { LabelContent }
