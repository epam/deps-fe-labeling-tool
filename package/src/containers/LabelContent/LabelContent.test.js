
import { mockUuid } from '@/mocks/mockUuid'
import React from 'react'
import { shallow } from 'enzyme'
import { Label, LabelType } from '@/models/Label'
import { LabelContent } from './LabelContent'
jest.mock('uuid', () => mockUuid)

describe('Container: LabelContent', () => {
  it('should render correct ContentTextArea layout based on props', () => {
    const label = new Label(
      0.113,
      0.8104,
      0.017,
      0.0119,
      'enclosures',
      undefined,
      LabelType.STRING,
      'hello world',
      { data: 'some meta for enclosures value label' },
      0.66
    )
    const wrapper = shallow(<LabelContent label={label} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render correct ContentCheckmarkRadio layout based on props', () => {
    const label = new Label(
      0.113,
      0.8104,
      0.017,
      0.0119,
      'enclosures',
      undefined,
      LabelType.CHECKMARK,
      true,
      { data: 'some meta for enclosures value label' },
      0.66
    )
    const wrapper = shallow(<LabelContent label={label} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render correct ContentEnum layout based on props', () => {
    const label = new Label(
      0.113,
      0.8104,
      0.017,
      0.0119,
      'enum',
      undefined,
      LabelType.ENUM,
      'test',
      { data: 'some meta for enum value label' },
      0.66
    )
    const wrapper = shallow(<LabelContent label={label} />)
    expect(wrapper).toMatchSnapshot()
  })
})
