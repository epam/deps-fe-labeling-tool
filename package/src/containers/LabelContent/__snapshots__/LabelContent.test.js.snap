// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: LabelContent should render correct ContentCheckmarkRadio layout based on props 1`] = `
<Connect(ContentCheckmarkRadio)
  label={
    Label {
      "confidence": 0.66,
      "content": true,
      "fieldCode": "enclosures",
      "h": 0.0119,
      "index": undefined,
      "meta": Object {
        "data": "some meta for enclosures value label",
      },
      "type": "checkmark",
      "typeName": "label",
      "uid": "2",
      "w": 0.017,
      "x": 0.113,
      "y": 0.8104,
    }
  }
/>
`;

exports[`Container: LabelContent should render correct ContentEnum layout based on props 1`] = `
<Connect(ContentEnum)
  label={
    Label {
      "confidence": 0.66,
      "content": "test",
      "fieldCode": "enum",
      "h": 0.0119,
      "index": undefined,
      "meta": Object {
        "data": "some meta for enum value label",
      },
      "type": "enum",
      "typeName": "label",
      "uid": "3",
      "w": 0.017,
      "x": 0.113,
      "y": 0.8104,
    }
  }
/>
`;

exports[`Container: LabelContent should render correct ContentTextArea layout based on props 1`] = `
<Connect(ContentTextArea)
  label={
    Label {
      "confidence": 0.66,
      "content": "hello world",
      "fieldCode": "enclosures",
      "h": 0.0119,
      "index": undefined,
      "meta": Object {
        "data": "some meta for enclosures value label",
      },
      "type": "string",
      "typeName": "label",
      "uid": "1",
      "w": 0.017,
      "x": 0.113,
      "y": 0.8104,
    }
  }
/>
`;
